Intro and outros slideshows, and dreams

Moon613 opened the pull request:
Unsure if the intro and possible outros are better kept separate as they are, or if it's best to merge the intro into the outro list, making it generic for both and giving the slideshows a bool to mark if they are an intro or not.
Also, right now the default music is forced to be the normal Survivor Intro music "RW_Intro_Theme", but I've been wondering if it should instead just be silent so that that's an option for users, since right now it is not a possibility.

Moon613 committed:
Support for intro cutscenes for custom slugcats

Moon613 committed:
Multiple images per scene

Moon613 committed:
Scene names can now be non-unique

Moon613 committed:
Able to add outro scenes now.
Also small helper function for calling them
Will potentially be expanded later

Moon613 committed:
Dream Scene Support.
Extra functions for slugcat creators to use in their code.
Changed some foreach loops into if statements.

Moon613 committed:
Dynamic Slideshow Movement.
Small bugfixes

Moon613 committed:
Default "credits" now sends you to credits.

Moon613 committed:
Reduce nesting directly in hooks.
Moved to methods for easier readability

Moon613 committed:
No More try catch for slideshow movements.

Moon613 committed:
Cleanup part 1

Moon613 committed:
Cleanup part 2

Moon613 committed:
Clearing the upcoming dream if it's irrelevant

Moon613 commented:
Meant to send this to my own main branch before making a PR so I could be sure that it was all cleaned up properly, but I goofed ¯\_(ツ)_/¯. It should be good now anyway, probably.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -67,5 +102,252 @@ private static void MenuScene_BuildScene(On.Menu.MenuScene.orig_BuildScene orig,
+                if (manager.oldProcess is Menu.SlugcatSelectMenu slugcatmenu) { name = slugcatmenu.slugcatPages[slugcatmenu.slugcatPageIndex].slugcatNumber; }
+                if (manager.oldProcess is RainWorldGame rainGame) { name = rainGame.StoryCharacter; }
+                if (SlugBaseCharacter.TryGet(name, out var chara)) {
+                    if (IntroScene.TryGet(chara, out var newIntroSlideShowID) && slideShowID == newIntroSlideShowID && CustomSlideshow.Registry.TryGet(newIntroSlideShowID, out var customIntroSlideshow))
```
`CustomSlideshow.Registry` should be the only thing that matters here. Even if the slugcat that owns the slideshow isn't currently selected, the slideshow should ideally still work. This current code will also not respect any slideshows that are added to the registry directly unless they're also added to an "intro_scene" or "outro_scenes" feature.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -67,5 +102,252 @@ private static void MenuScene_BuildScene(On.Menu.MenuScene.orig_BuildScene orig,
+            }
+            if (SlugBaseCharacter.TryGet(name, out var chara))
+            {
+                if (IntroScene.TryGet(chara, out var newIntroSlideShowID) && self.menu is SlideShow slideShow && slideShow.slideShowID == newIntroSlideShowID && CustomSlideshow.Registry.TryGet(newIntroSlideShowID, out var customIntroSlideshow))
```
As with `SlideShow_ctor`, this should forgo checking any of the slugcat-specific features and only check the slideshow registry.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -67,5 +102,252 @@ private static void MenuScene_BuildScene(On.Menu.MenuScene.orig_BuildScene orig,
+                name = rainGame.StoryCharacter;
+            }
+            if (SlugBaseCharacter.TryGet(name, out var chara)) {
+                if (IntroScene.TryGet(chara, out var newIntroSlideShowID) && self.menu is SlideShow slideShow && slideShow.slideShowID == newIntroSlideShowID && CustomSlideshow.Registry.TryGet(newIntroSlideShowID, out var customIntroSlideshow))
```
As with `SlideShow_ctor`, this should forgo checking any of the slugcat-specific features and only check the slideshow registry.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -67,5 +102,252 @@ private static void MenuScene_BuildScene(On.Menu.MenuScene.orig_BuildScene orig,
+            SceneID origSceneID = orig(self, dreamID);
+            if (self.manager.oldProcess is RainWorldGame rainGame && SlugBaseCharacter.TryGet(rainGame.StoryCharacter, out var chara) && HasDreams.TryGet(chara, out bool dreams) && dreamID.value == CustomScene.nextDreamID)
+            {
+                CustomScene.QueueDream("");
```
Try not to add side-effects to methods that didn't previously have them. I recommend reworking the dream system to not rely on static variables.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -67,5 +102,252 @@ private static void MenuScene_BuildScene(On.Menu.MenuScene.orig_BuildScene orig,
+            foreach (var scene in customSlideshow.Scenes) {
+                self.playList.Add(new Scene(new SceneID( $"Slugbase_{charaName}_{customSlideshow.ID}_{scene.Name}_intro", false), self.ConvertTime(0, scene.StartAt, 0), self.ConvertTime(0, scene.FadeInDoneAt, 0), self.ConvertTime(0, scene.FadeOutStartAt, 0)));
+            }
+            if (intro) { self.processAfterSlideShow = ProcessManager.ProcessID.Game; }
```
Would it be better if the slideshow JSON took "next_process" instead of "to_credits"? I'm not sure.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -67,5 +102,252 @@ private static void MenuScene_BuildScene(On.Menu.MenuScene.orig_BuildScene orig,
+        {
+            foreach (var scene in customSlideshow.Scenes)
+            {
+                if (new SceneID($"Slugbase_{charaName}_{customSlideshow.ID}_{scene.Name}_intro", false) == sceneID && !self.flatMode)
```
Consider adding a method, `CustomSlideshow.GetScene(string name)`, that does this search for you. Storing a `Dictionary<string, Scene>` would make it more convenient and skip a bunch of string interpolation.

SlimeCubed commented on src/Assets/CustomScene.cs:
```diff
@@ -13,6 +13,20 @@ namespace SlugBase.Assets
+        /// Set the dream scene that will display when the player hibernates next.
+        /// </summary>
+        /// <param name="name">The id of the scene to display.</param>
+        public static void QueueDream(string name)
```
This should be reworked to not depend on static variables. In its current state, it's rather fragile. Consider having the modder pass in a `DreamsState`, then either store the queued dream with a CWT or somehow use the existing `eventDream` system. This should make the dream leaking between save states impossible

SlimeCubed commented on src/Assets/CustomScene.cs:
```diff
@@ -13,6 +13,20 @@ namespace SlugBase.Assets
+        /// <summary>
+        /// The ID of the next custom dream to play, as a string
+        /// </summary>
+        public static string nextDreamID { get; private set; } = "";
```
This should be capitalized since it's a property.

SlimeCubed commented on src/Assets/CustomSlideshow.cs:
```diff
@@ -0,0 +1,239 @@
+        /// <summary>
+        /// A scene from a <see cref="CustomSlideshow"/> that holds data about when to appear and what images to use for what amount of time
+        /// </summary>
+        public class Scene
```
I don't like how much of this is copy-pasted from `CustomScene`. You should remove the duplicate properties, then either make this class inherit from `CustomScene` or add a property that stores a `CustomScene`.

SlimeCubed commented on src/Assets/CustomSlideshow.cs:
```diff
@@ -0,0 +1,239 @@
+        /// <summary>
+        /// Data about a song from a <see cref="CustomSlideshow"/>.
+        /// </summary>
+        public class MMusic{
```
Maybe call this `SlideshowMusic` instead of `MMusic`? I'd rather be a little redundant than have people wonder what the "M" stands for.

SlimeCubed commented on src/Assets/CustomSlideshow.cs:
```diff
@@ -0,0 +1,239 @@
+                .ToArray();
+
+            SlideshowFolder = json.TryGet("slideshow_folder")?.AsString().Replace('/', Path.DirectorySeparatorChar);
+            // Don't know if I should force it to defalut to the normal intro theme or leave it empty so that it's an option for people to not have any music (But who would choose that? Someone probably)
```
I'd rather it default to silence. The slideshow system should be fairly flexible, so who knows what people will use it for. Maybe they want to handle the music logic themselves!

SlimeCubed commented on src/Features/Features.cs:
```diff
@@ -147,6 +147,17 @@ public static class GameFeatures
+            var list = json.AsList();
+            return list.Select(JsonUtils.ToExtEnum<SlideShow.SlideShowID>).ToArray();
+        });
+        // This ideally should be able to support an arbitrary amount of outros for alternate endings. Personally I'd also be fine with only allowing two, one being `outro_scene_alt`, seems like a reasonable compromise
```
Any slideshow in the `slugbase\slideshows` should be automatically loaded, meaning this feature isn't actually that important; it serves no purpose except indicating which one should be used as the default outro scene.
I think this should be renamed to `"outro_slideshow"` and accept a single slideshow ID. If the modder wants to add alternate endings, they will need to write C# anyways and should add their own feature.

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -13,10 +19,39 @@ internal static class AssetHooks
+        private static void Player_Die(On.Player.orig_Die orig, Player self)
+        {
+            orig(self);
+            CustomScene.QueueDream("");
```
What if the player dies but is revived, a player in coop dies, or a slugpup dies?

SlimeCubed commented on src/Assets/AssetHooks.cs:
```diff
@@ -13,10 +19,39 @@ internal static class AssetHooks
+        private static void MainMenu_ctor(On.Menu.MainMenu.orig_ctor orig, MainMenu self, ProcessManager manager, bool showRegionSpecificBkg)
+        {
+            orig(self, manager, showRegionSpecificBkg);
+            CustomScene.QueueDream("");
```
What if a mod causes the game to exit to a process other than the main menu?

SlimeCubed reviewed (changes requested):
Thanks for your contribution! I wasn't looking forward to making this bit, so I'm glad someone else got to it before me :)

Most of it looks good, but there are a few things that I do think should be changed. In general, the slideshow system currently has too many special cases where it only works if the slideshow is an intro or outro. Leave a comment here if you need any help/info/whatever, but I'll probably see messages sooner on Discord.

Moon613 committed:
Simplified Slideshow Logic

Moon613 committed:
Revert "Simplified Slideshow Logic"

This reverts commit e1a5e0e6f378a21c89af253d43a11c7ce02f1e68.

Moon613 committed:
Revert "Revert "Simplified Slideshow Logic""

This reverts commit 7cb187160a7229d7fdce5f93cb8bef9f759c9acc.

Moon613 committed:
Fix/Overhaul Dream Code.

Moon613 committed:
Add GetScene method to CustomSlideshow class.

Moon613 committed:
Clean-up.

Moon613 committed:
Change the Select Scene if you use an alt outro.

Moon613 committed:
Update Docs

Moon613 committed:
Fix/Finish alt ending select screen.

Now it doesn't replace all slugbase scugs, and properly stays across game restarts.

Moon613 committed:
Make sure a menu scene exists before using it.

Moon613 committed:
Very Minor Cleanup

Moon613 committed:
Don't save the room ended in. Only Misc data.

Moon613 committed:
Jolly color support for scugs with up to 9 colors.

SlimeCubed committed:
Merge Jolly changes

SlimeCubed committed:
Clean up and refactor

SlimeCubed committed:
Fix save data and save data mining
